sites/partners/src/applications/PaperApplicationForm/sections/FormPreferences.tsx

```diff
@@ -1,66 +1,130 @@
+      preference?.formMetadata.options.forEach((option) =>
+        keys.push(buildOptionName(preference?.formMetadata.key, option.key))
+      )
+    )
+
+    return keys
+  }, [preferences, buildOptionName])
 
-            <Field
-              id="application.preferences.workIn"
-              name="application.preferences.workIn"
-              type="checkbox"
-              label={`${t("application.add.preferences.workIn")} ${t(
-                "application.details.countyName"
-              )}`}
-              register={register}
-              inputProps={{
-                onChange: () => {
-                  setValue("application.preferences.none", false)
-                },
-              }}
-            />
+  const watchPreferences = watch(allOptionFieldNames)
 
-            <Field
-              id="application.preferences.none"
-              name="application.preferences.none"
-              type="checkbox"
-              label={t("application.add.preferences.optedOut")}
-              register={register}
-              inputProps={{
-                onChange: () => {
-                  setValue("application.preferences.liveIn", false)
-                  setValue("application.preferences.workIn", false)
-                },
-              }}
-            />
-          </fieldset>
-        </ViewItem>
-      </GridCell>
+  function uncheckPreference(metaKey: string, options: FormMetadataOptions[]) {
+    const preferenceKeys = options?.map((option) => option.key)
+    preferenceKeys.forEach((k) => setValue(buildOptionName(metaKey, k), false))
+  }
+
+  const createExtraDataFields = useCallback(
+    (metaKey: string, optionKey: string, extraKey: string, type: InputType) => {
+      const FIELD_NAME = `${PREFERENCES_FORM_PATH}.${metaKey}.${optionKey}.${extraKey}`
+
+      return (
+        <div className="my-4" key={FIELD_NAME}>
+          {(() => {
+            if (type === InputType.text) {
+              return (
+                <Field
+                  id={FIELD_NAME}
+                  name={FIELD_NAME}
+                  type="text"
+                  label={t(`application.preferences.options.${extraKey}`)}
+                  register={register}
+                />
+              )
+            }
+
+            if (type === InputType.address) {
+              return FormAddress(
+                t("application.contact.address"),
+                FIELD_NAME,
+                "preference",
+                register
+              )
+            }
+          })()}
+        </div>
+      )
+    },
+    [register]
+  )
+
+  return (
+    <GridSection title={t("application.details.preferences")} separator grid={false}>
+      <GridSection columns={2}>
+        {preferences?.map((preference) => (
+          <GridCell key={preference.id}>
+            <ViewItem label={preference.title}>
+              <fieldset className="mt-4">
+                {preference.formMetadata?.options?.map((option) => {
```

Comment:
Could this section and the memo/callback blocks above be refactored into regular functional components? It might clean up some of the procedural nature of the code here.

Reply:
Definitely makes sense! 👍  I think we can store this component in the same file (logic is highly connected to this component).